File: util/persistence/configuration/src/main/java/dev/sweety/config/json/GsonUtils.java
package dev.sweety.config.json;

import com.google.gson.Gson;

import java.io.Reader;
import java.lang.reflect.Type;

public final class GsonUtils {

    private static final ThreadLocal<Gson> gson = ThreadLocal.withInitial(() -> new Gson().newBuilder().disableHtmlEscaping().setPrettyPrinting().create());

    public static Gson gson() {
        return gson.get();
    }

    public static <T> String write(T obj) {
        return gson().toJson(obj);
    }

    public static <T> String write(T obj, Type type) {
        return gson().toJson(obj, type);
    }

    public static <T> void save(T config, Appendable writer) {
        gson().toJson(config, writer);
    }

    public static <T> void save(T config, Type type, Appendable writer) {
        gson().toJson(config, type, writer);
    }

    public static <T> T load(Reader reader, Class<T> configClass) {
        return gson().fromJson(reader, configClass);
    }

    public static <T> T read(String obj, Class<T> clazz) {
        return gson().fromJson(obj, clazz);
    }

    public static <T> T load(Reader reader, Type configClass) {
        return gson().fromJson(reader, configClass);
    }

    public static <T> T read(String obj, Type clazz) {
        return gson().fromJson(obj, clazz);
    }

    private GsonUtils() {

    }

}

File: util/persistence/configuration/src/main/kotlin/dev/sweety/config/json/GsonUtils.kt
package dev.sweety.config.json

import com.google.gson.Gson
import java.io.Reader
import java.lang.reflect.Type

object GsonUtils {

    private val gson: ThreadLocal<Gson> =
        ThreadLocal.withInitial { Gson().newBuilder().disableHtmlEscaping().setPrettyPrinting().create() }

    @JvmStatic
    fun gson(): Gson = gson.get()

    @JvmStatic
    fun <T> write(obj: T): String = gson().toJson(obj)

    @JvmStatic
    fun <T> write(obj: T, type: Type): String = gson().toJson(obj, type)

    @JvmStatic
    fun <T> save(config: T, writer: Appendable) {
        gson().toJson(config, writer)
    }

    @JvmStatic
    fun <T> save(config: T, type: Type, writer: Appendable) {
        gson().toJson(config, type, writer)
    }

    @JvmStatic
    fun <T> load(reader: Reader, configClass: Class<T>): T = gson().fromJson(reader, configClass)

    @JvmStatic
    fun <T> read(obj: String, clazz: Class<T>): T = gson().fromJson(obj, clazz)

    @JvmStatic
    fun <T> load(reader: Reader, configClass: Type): T = gson().fromJson(reader, configClass)

    @JvmStatic
    fun <T> read(obj: String, clazz: Type): T = gson().fromJson(obj, clazz)
}
